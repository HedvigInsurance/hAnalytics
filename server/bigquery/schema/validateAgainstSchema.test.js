const validateAgainstSchema = require("./validateAgainstSchema");
const setupTable = require("./setupTable");
const createBigQueryConfigMock = require("../config.mock");
const { bigQuerySchemaTypeMap } = require("../../../commons/typeMaps");

test("validate String", async () => {
  const bigQueryConfig = createBigQueryConfigMock();

  await setupTable(
    "mock_table",
    "mock table",
    [
      {
        name: "property_hello",
        ...bigQuerySchemaTypeMap("String"),
      },
    ],
    bigQueryConfig
  );

  const validate = await validateAgainstSchema(
    "mock_table",
    {
      property_hello: "hello",
    },
    bigQueryConfig
  );

  expect(validate).toEqual(true);
});

test("validate should fail with unknown properites", async () => {
  const bigQueryConfig = createBigQueryConfigMock();

  await setupTable(
    "mock_table",
    "mock table",
    [
      {
        name: "property_hello",
        ...bigQuerySchemaTypeMap("String"),
      },
    ],
    bigQueryConfig
  );

  const validate = await validateAgainstSchema(
    "mock_table",
    {
      property_not_in_schema: "mock",
    },
    bigQueryConfig
  );

  expect(validate).toEqual(false);
});

test("validate should fail with wrong type", async () => {
  const bigQueryConfig = createBigQueryConfigMock();

  await setupTable(
    "mock_table",
    "mock table",
    [
      {
        name: "property_hello",
        ...bigQuerySchemaTypeMap("String"),
      },
    ],
    bigQueryConfig
  );

  const validate = await validateAgainstSchema(
    "mock_table",
    {
      property_hello: 123,
    },
    bigQueryConfig
  );

  expect(validate).toEqual(false);
});

test("validate should accept nullable", async () => {
  const bigQueryConfig = createBigQueryConfigMock();

  await setupTable(
    "mock_table",
    "mock table",
    [
      {
        name: "property_mock",
        ...bigQuerySchemaTypeMap("Optional<String>"),
      },
    ],
    bigQueryConfig
  );

  const validate = await validateAgainstSchema(
    "mock_table",
    {
      property_mock: null,
    },
    bigQueryConfig
  );

  expect(validate).toEqual(true);
});

test("validate should not accept null", async () => {
  const bigQueryConfig = createBigQueryConfigMock();

  await setupTable(
    "mock_table",
    "mock table",
    [
      {
        name: "property_hello",
        ...bigQuerySchemaTypeMap("String"),
      },
    ],
    bigQueryConfig
  );

  const validate = await validateAgainstSchema(
    "mock_table",
    {
      property_hello: null,
    },
    bigQueryConfig
  );

  expect(validate).toEqual(false);
});

test("validate should accept array", async () => {
  const bigQueryConfig = createBigQueryConfigMock();

  await setupTable(
    "mock_table",
    "mock table",
    [
      {
        name: "experiments",
        ...bigQuerySchemaTypeMap("Array<String>"),
      },
    ],
    bigQueryConfig
  );

  const validate = await validateAgainstSchema(
    "mock_table",
    {
      experiments: ["123", "456"],
    },
    bigQueryConfig
  );

  expect(validate).toEqual(true);
});

test("validate should not accept array with wrong type", async () => {
  const bigQueryConfig = createBigQueryConfigMock();

  await setupTable(
    "mock_table",
    "mock table",
    [
      {
        name: "experiments",
        ...bigQuerySchemaTypeMap("Array<String>"),
      },
    ],
    bigQueryConfig
  );

  const validate = await validateAgainstSchema(
    "mock_table",
    {
      experiments: [123, "456"],
    },
    bigQueryConfig
  );

  expect(validate).toEqual(false);
});

test("validate custom enum type values", async () => {
  const bigQueryConfig = createBigQueryConfigMock();

  await setupTable(
    "mock_table",
    "mock table",
    [
      {
        name: "app_screen_name",
        ...bigQuerySchemaTypeMap("AppScreen"),
      },
    ],
    bigQueryConfig
  );

  expect(bigQueryConfig.bigquery.getTables()).toMatchSnapshot();

  const validate = await validateAgainstSchema(
    "mock_table",
    {
      app_screen_name: "offer",
    },
    bigQueryConfig
  );

  expect(validate).toEqual(true);

  const validateFaulty = await validateAgainstSchema(
    "mock_table",
    {
      app_screen_name: "this_should_not_work",
    },
    bigQueryConfig
  );

  expect(validateFaulty).toEqual(false);
});
