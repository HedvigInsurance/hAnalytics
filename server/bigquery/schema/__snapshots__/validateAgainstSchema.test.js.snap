// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate custom enum type values 1`] = `
Array [
  Object {
    "config": Object {
      "description": "mock table",
      "schema": Object {
        "fields": Array [
          Object {
            "mode": "REQUIRED",
            "name": "app_screen_name",
            "permittedValues": Array [
              "offer",
              "claims_status_detail",
              "claim_honor_pledge",
              "common_claim_detail",
              "embark",
              "embark_tooltip",
              "market_picker",
              "marketing",
              "data_collection_authenticating",
              "data_collection_credentials",
              "data_collection_fail",
              "data_collection_intro",
              "data_collection_success",
              "connect_payment_adyen",
              "connect_payment_trustly",
              "connect_payment_failed",
              "connect_payment_success",
              "payments",
              "charity",
              "contact_info",
              "cross_sell_detail",
              "forever",
              "home",
              "insurance_Detail",
              "insurances",
              "moving_flow_intro",
              "profile",
              "app_information",
              "app_settings",
              "chat",
            ],
            "type": "STRING",
          },
        ],
      },
      "timePartitioning": Object {
        "expirationMS": "7776000000",
        "type": "DAY",
      },
    },
    "name": "mock_table",
    "rows": Array [],
  },
]
`;
